wooting-macro-backend/src/config.rs

```diff
@@ -1,117 +1,182 @@
 use crate::MacroData;
+use anyhow::Result;
 use std::fs::File;
 use std::path::PathBuf;
 
 use log::*;
 
 /// Trait to get data or write out data from the state to file.
 pub trait ConfigFile: Default + serde::Serialize + for<'de> serde::Deserialize<'de> {
-    fn file_name() -> PathBuf;
+    fn file_name() -> Result<PathBuf>;
 
     /// Reads the data from the file and returns it.
     ///
     /// If it errors out, it replaces and writes a default config.
-    fn read_data() -> Self {
-        let default = Self::default();
-
-        match File::open(Self::file_name().as_path()) {
+    fn read_data() -> Result<Self> {
+        match File::open(Self::file_name()?) {
             Ok(data) => {
                 let data: Self = match serde_json::from_reader(&data) {
                     Ok(x) => x,
                     Err(error) => {
-                        error!("Error reading config.json, using default data. {}", error);
-                        default.write_to_file();
-                        default
+                        error!("Error reading config.json, using default data and moving to backup. {}", error);
+                        let file_name = Self::file_name();
+                        let mut file_name = file_name?
+                            .to_str()
+                            .expect("Cannot preserve invalid file, crashing to save custom config")
+                            .to_string();
+
+                        if PathBuf::from(format!("{}.bak", file_name)).exists() {
+                            let mut i = 0;
+                            while PathBuf::from(format!("{}-{}.bak", file_name, i)).exists() {
+                                i += 1;
+                            }
+                            file_name = format!("{}-{}.bak", file_name, i);
+                        }
+                        warn!(
+                            "Renaming {} to {}",
+                            Self::file_name()?.to_str().unwrap(),
+                            file_name
+                        );
+
+                        std::fs::rename(Self::file_name()?, file_name)
+                            .expect("cannot rename the invalid config, aborting.");
+                        Self::default()
```

Comment:
more robust handling of overwriting files but this needs to be also changed in 1.2 in frontend due to the migration system needing to be done.

Reply:
Adding that this entire code will definitely undergo a redesign in 1.2, as I'm really not happy with how it works overall.